go/merkletree_new/merkle_tree_interface.go

```diff
@@ -0,0 +1,65 @@
+package merkletree
+
+// MerkleTreeInterface is the methods which are required to manipulate the
+// contents of a Merkle Hash Tree (that is, add new entries), and to query
+// the tree for inclusion and transparency proofs.
+type MerkleTreeInterface interface {
+	// CurrentRootHash returns the hash of the entire current tree.
+	CurrentRootHash func () ([]byte, error)
+
+	// InclusionProof returns a list of the hashes of the "sibling"
+	// nodes to the leaf entry located at the (0-based) index specified.
+	// The returned list is presented in order, starting from the
+	// sibling to the item itself, and ending with a hash that is an
+	// immediate child of the root.
+	InclusionProof func (entry uint64) ([]byte, error)
+
+	// ConsistencyProof returns a list of the hashes that make up the
+	// consistency proof between the hash trees with size |tree1| and
+	// |tree2|.
+	ConsistencyProof func (sz uint64) ([]byte, error)
+}
+
+// MerkleTreeDataInterface defines the means by which a merkle hash tree can
+// retrieve data from an arbitrary data store, without tying the tree
+// implementation to any particular backend storage technology.
+type MerkleTreeDataInterface interface {
+	// EntryAt takes a (0-based) |index| into the list of leaf entries in
+	// the tree, and returns the contents of the leaf.  For a given
+	// |index| in a given tree, the value returned **MUST NOT** ever
+	// change, or the whole point of a merkle tree is defeated.
+	EntryAt func (index uint64) []byte
+
+	// Size returns the number of entries currently in the tree.  The
+	// value returned by this function must never decrease between
+	// calls; it is OK for the value to increase over time.
+	Size func () uint64
+}
+
+// MerkleTreeCacheInterface provides a means for a merkle tree to cache
+// hash values in the tree.  Whilst all hash values *can* be recalculated
+// on the fly, it is far more efficient to be able to cache recently-used
+// values, to avoid needing to rehash everything.
+type MerkleTreeCacheInterface interface {
+	GetNode func ([]byte) []byte
+	SetNode func ([]byte) []byte
+}
+
+type MerkleTree struct {
+	// If a merkle tree you wish to be, you must implement these
+	// functions three.
+	MerkleTreeInterface
+
+	dao   *MerkleTreeDataInterface
+	cache *MerkleTreeCacheInterface
```

Comment:
That would require the cache interface to know an awful lot about how to construct intermediate nodes -- essentially, half of the merkle tree implementation would end up inside the cache interface.  My desire is that the cache interface be "pluggable".  The first cache interface would be a simple in-memory LRU cache, however a shared cache (using, say, Redis or memcached) would be far more CPU efficient and allow new nodes to be immediately effective, rather than having to pre-calculate the entire tree.